### PR TITLE
Add support for Gradle 8

### PR DESCRIPTION
**Pull request by snoopcheri**
This PR adds compatibility with the recent changes in Gradle 8 nightly.

It is based on the changes in https://github.com/gradle/gradle/pull/22139 and shows how those changes would be used in the test retry plugin.

Relates to https://github.com/gradle/test-retry-gradle-plugin/issues/148

### CODE REVIEW

File: plugin/src/main/java/org/gradle/testretry/internal/executer/framework/Junit5TestFrameworkStrategy.java [hunk 1]
```diff
@@ -16,24 +16,55 @@
...
+    }
+
+    private static JUnitPlatformTestFramework newJUnitPlatformTestFrameworkInstanceForGradleOlderThanV8_0(Test task, DefaultTestFilter failedTestsFilter) {
+        try {
+            Class<?> jUnitPlatformTestFrameworkClass = JUnitPlatformTestFramework.class;
+            Constructor<?> constructor = jUnitPlatformTestFrameworkClass.getConstructor(DefaultTestFilter.class);
+
+            return (JUnitPlatformTestFramework) constructor.newInstance(failedTestsFilter);
+        } catch (NoSuchMethodException | InstantiationException | IllegalAccessException |
+                 InvocationTargetException e) {
```
Comment (marcphilipp) [inline plugin/src/main/java/org/gradle/testretry/internal/executer/framework/Junit5TestFrameworkStrategy.java:62]:
```suggestion
        } catch (ReflectiveOperationException e) {
```

**Review by marcphilipp — APPROVED** [review]

File: plugin/src/main/java/org/gradle/testretry/internal/executer/framework/Junit5TestFrameworkStrategy.java (hunk 1 above)
Comment (snoopcheri) [reply plugin/src/main/java/org/gradle/testretry/internal/executer/framework/Junit5TestFrameworkStrategy.java]: DONE (26701fe8)

File: plugin/src/main/java/org/gradle/testretry/internal/executer/framework/Junit5TestFrameworkStrategy.java [hunk 2]
```diff
@@ -16,27 +16,82 @@
...
+            }
+
+            private static JUnitPlatformTestFramework newInstance(DefaultTestFilter failedTestsFilter) {
+                try {
+                    Class<?> jUnitPlatformTestFrameworkClass = JUnitPlatformTestFramework.class;
+                    Constructor<?> constructor = jUnitPlatformTestFrameworkClass.getConstructor(DefaultTestFilter.class);
+
+                    return (JUnitPlatformTestFramework) constructor.newInstance(failedTestsFilter);
+                } catch (ReflectiveOperationException e) {
+                    throw new RuntimeException(e);
```
Comment (leonard84) [inline plugin/src/main/java/org/gradle/testretry/internal/executer/framework/Junit5TestFrameworkStrategy.java:75]: Don't we have `UncheckedException` for this, or some other more fitting type?

File: plugin/src/main/java/org/gradle/testretry/internal/executer/framework/Junit5TestFrameworkStrategy.java [hunk 3]
```diff
@@ -16,27 +16,82 @@
...
+            private final TestFramework testFramework;
+
+            ProviderForCurrentGradleVersion(TestFramework testFramework) {
+                this.testFramework = testFramework;
+            }
+
+            @Override
+            public TestFramework testFrameworkFor(DefaultTestFilter failedTestsFilter) {
+                return testFramework.copyWithFilters(failedTestsFilter);            }
+        }
```
Comment (leonard84) [inline plugin/src/main/java/org/gradle/testretry/internal/executer/framework/Junit5TestFrameworkStrategy.java:50]: ❓ Why can't we share this implementation, it is the same across all frameworks.

**Review by leonard84 — CHANGES_REQUESTED** [review]

File: plugin/src/main/java/org/gradle/testretry/internal/executer/framework/Junit5TestFrameworkStrategy.java (hunk 3 above)
Comment (marcphilipp) [reply plugin/src/main/java/org/gradle/testretry/internal/executer/framework/Junit5TestFrameworkStrategy.java:50]: I was also wondering that.

File: plugin/src/main/java/org/gradle/testretry/internal/executer/framework/Junit5TestFrameworkStrategy.java [hunk 4]
```diff
@@ -16,27 +16,82 @@
...
+
+            private final TestFramework testFramework;
+
+            ProviderForCurrentGradleVersion(TestFramework testFramework) {
+                this.testFramework = testFramework;
+            }
+
+            @Override
+            public TestFramework testFrameworkFor(DefaultTestFilter failedTestsFilter) {
+                return testFramework.copyWithFilters(failedTestsFilter);            }
```
Comment (marcphilipp) [inline plugin/src/main/java/org/gradle/testretry/internal/executer/framework/Junit5TestFrameworkStrategy.java:49]:
```suggestion
                return testFramework.copyWithFilters(failedTestsFilter);
            }
```

File: plugin/src/main/java/org/gradle/testretry/internal/executer/framework/Junit5TestFrameworkStrategy.java (hunk 2 above)
Comment (snoopcheri) [reply plugin/src/main/java/org/gradle/testretry/internal/executer/framework/Junit5TestFrameworkStrategy.java:75]: I've used the same exception as we're currently throwing in similar situations like the [RetryTestResultProcessor](https://github.com/gradle/test-retry-gradle-plugin/blob/main/plugin/src/main/java/org/gradle/testretry/internal/executer/RetryTestResultProcessor.java#L139-L141).

I think it makes sense to always throw the same here and therefore I suggest to keep it that way in order to minimize risk that we're breaking something else with customers. Personally, I probably would have chosen `UnsupportedOperationException`.

@leonard84 WDYT?

File: plugin/src/main/java/org/gradle/testretry/internal/executer/framework/Junit5TestFrameworkStrategy.java (hunk 3 above)
Comment (snoopcheri) [reply plugin/src/main/java/org/gradle/testretry/internal/executer/framework/Junit5TestFrameworkStrategy.java]: Good idea.

DONE (45ed11e2)

**Review by jvandort — APPROVED** [review]

**Review by leonard84 — APPROVED** [review]

**Pull request merged**